lib/css/CssModulesPlugin.js

```diff
@@ -0,0 +1,246 @@
+					new StaticExportsDependency.Template()
+				);
+				normalModuleFactory.hooks.createParser
+					.for("css")
+					.tap(plugin, parserOptions => {
+						validateParserOptions(parserOptions);
+						return new CssParser();
+					});
+				normalModuleFactory.hooks.createGenerator
+					.for("css")
+					.tap(plugin, generatorOptions => {
+						validateGeneratorOptions(generatorOptions);
+						return new CssGenerator();
+					});
+				compilation.hooks.contentHash.tap("JavascriptModulesPlugin", chunk => {
+					const {
+						chunkGraph,
+						outputOptions: {
+							hashSalt,
+							hashDigest,
+							hashDigestLength,
+							hashFunction
+						}
+					} = compilation;
+					if (!CssModulesPlugin.chunkHasCss(chunk, chunkGraph)) return;
+					const hash = createHash(hashFunction);
+					if (hashSalt) hash.update(hashSalt);
+					const modules = this.getOrderedChunkCssModules(chunk, chunkGraph);
+					for (const module of modules) {
+						hash.update(chunkGraph.getModuleHash(module, chunk.runtime));
+					}
+					const digest = /** @type {string} */ (hash.digest(hashDigest));
+					chunk.contentHash.css = digest.substr(0, hashDigestLength);
+				});
+				compilation.hooks.renderManifest.tap(plugin, (result, options) => {
+					const { chunkGraph } = compilation;
+					const { hash, chunk, codeGenerationResults } = options;
+
+					if (chunk instanceof HotUpdateChunk) return result;
+
+					if (CssModulesPlugin.chunkHasCss(chunk, chunkGraph)) {
+						result.push({
+							render: () =>
+								this.renderChunk({
+									chunk,
+									chunkGraph,
+									codeGenerationResults
+								}),
+							filenameTemplate: CssModulesPlugin.getChunkFilenameTemplate(
+								chunk,
+								compilation.outputOptions
+							),
+							pathOptions: {
+								hash,
+								runtime: chunk.runtime,
+								chunk,
+								contentHashType: "css"
+							},
+							identifier: `css${chunk.id}`,
+							hash: chunk.contentHash.css
+						});
+					}
+					return result;
+				});
+				const enabledChunks = new WeakSet();
+				const handler = (chunk, set) => {
+					if (enabledChunks.has(chunk)) {
+						return;
+					}
+					enabledChunks.add(chunk);
+
+					set.add(RuntimeGlobals.publicPath);
+					set.add(RuntimeGlobals.getChunkCssFilename);
+					set.add(RuntimeGlobals.hasOwnProperty);
+					set.add(RuntimeGlobals.moduleFactoriesAddOnly);
+					set.add(RuntimeGlobals.makeNamespaceObject);
+
+					const CssLoadingRuntimeModule = getCssLoadingRuntimeModule();
+					compilation.addRuntimeModule(chunk, new CssLoadingRuntimeModule(set));
+				};
+				compilation.hooks.runtimeRequirementInTree
+					.for(RuntimeGlobals.ensureChunkHandlers)
+					.tap(plugin, handler);
+				compilation.hooks.runtimeRequirementInTree
+					.for(RuntimeGlobals.hmrDownloadUpdateHandlers)
+					.tap(plugin, handler);
+			}
+		);
+	}
+
+	getOrderedChunkCssModules(chunk, chunkGraph) {
+		const cssImports = chunkGraph.getOrderedChunkModulesIterableBySourceType(
```

Comment:
probably `css-import` and `css` should be in one "order" call. e.g. in code below
```
import "https://h.ost/a.css" // css-import
import "./a.css"
```
hash order should be ['css-import', 'css']
and next one
```
import "./a.css"
import "https://h.ost/a.css" // css-import
```
hash order should be ['css', 'css-import']

maybe `getOrderedChunkModulesIterableBySourceType ` could be improved to allow pass `Set` as argument

Reply:
In a CSS file all `@import` must be before normal rules. So this enforces this order.